<a href="https://colab.research.google.com/github/vijayasri-manda/Data-Science/blob/main/NLTK_and_SPACY.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **NLP:**Natural Language Processing
NLP is not a language or library it is a field of AI,Which is used for analyzing and understanding **Human language** through python libaries such as **NLTK,SPACY,GENSIM** and etc..


# Applications of NLP
* spell Checking
* keyword Search
* Information Extraction
* Advertisement Matching
* Sentimental Analysis
* Speech Recognition
* Machine Translation

# **NLTK:**Natural Lanaguage Tool Kit
It is a python library for performing various natural language processing tasks.

**Some Crucial NLP Tasks:**
*   Tokenization
*   Stemming
*   Lematization






In [1]:
import nltk
from nltk.corpus import stopwords
nltk.download('stopwords')


[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


True

**Stopwords** means commonly used words

In [2]:
nltk_stopwords=nltk.corpus.stopwords.words('english')
nltk_stopwords

['a',
 'about',
 'above',
 'after',
 'again',
 'against',
 'ain',
 'all',
 'am',
 'an',
 'and',
 'any',
 'are',
 'aren',
 "aren't",
 'as',
 'at',
 'be',
 'because',
 'been',
 'before',
 'being',
 'below',
 'between',
 'both',
 'but',
 'by',
 'can',
 'couldn',
 "couldn't",
 'd',
 'did',
 'didn',
 "didn't",
 'do',
 'does',
 'doesn',
 "doesn't",
 'doing',
 'don',
 "don't",
 'down',
 'during',
 'each',
 'few',
 'for',
 'from',
 'further',
 'had',
 'hadn',
 "hadn't",
 'has',
 'hasn',
 "hasn't",
 'have',
 'haven',
 "haven't",
 'having',
 'he',
 "he'd",
 "he'll",
 'her',
 'here',
 'hers',
 'herself',
 "he's",
 'him',
 'himself',
 'his',
 'how',
 'i',
 "i'd",
 'if',
 "i'll",
 "i'm",
 'in',
 'into',
 'is',
 'isn',
 "isn't",
 'it',
 "it'd",
 "it'll",
 "it's",
 'its',
 'itself',
 "i've",
 'just',
 'll',
 'm',
 'ma',
 'me',
 'mightn',
 "mightn't",
 'more',
 'most',
 'mustn',
 "mustn't",
 'my',
 'myself',
 'needn',
 "needn't",
 'no',
 'nor',
 'not',
 'now',
 'o',
 'of',
 'off',
 'on',
 'once',
 'on

In [3]:
print("List of stop words in english:\n%s"%list(nltk_stopwords))

List of stop words in english:
['a', 'about', 'above', 'after', 'again', 'against', 'ain', 'all', 'am', 'an', 'and', 'any', 'are', 'aren', "aren't", 'as', 'at', 'be', 'because', 'been', 'before', 'being', 'below', 'between', 'both', 'but', 'by', 'can', 'couldn', "couldn't", 'd', 'did', 'didn', "didn't", 'do', 'does', 'doesn', "doesn't", 'doing', 'don', "don't", 'down', 'during', 'each', 'few', 'for', 'from', 'further', 'had', 'hadn', "hadn't", 'has', 'hasn', "hasn't", 'have', 'haven', "haven't", 'having', 'he', "he'd", "he'll", 'her', 'here', 'hers', 'herself', "he's", 'him', 'himself', 'his', 'how', 'i', "i'd", 'if', "i'll", "i'm", 'in', 'into', 'is', 'isn', "isn't", 'it', "it'd", "it'll", "it's", 'its', 'itself', "i've", 'just', 'll', 'm', 'ma', 'me', 'mightn', "mightn't", 'more', 'most', 'mustn', "mustn't", 'my', 'myself', 'needn', "needn't", 'no', 'nor', 'not', 'now', 'o', 'of', 'off', 'on', 'once', 'only', 'or', 'other', 'our', 'ours', 'ourselves', 'out', 'over', 'own', 're', 's',

In [4]:
print("length of stop words",len(nltk_stopwords))

length of stop words 198


In [5]:
print(stopwords.fileids())

['albanian', 'arabic', 'azerbaijani', 'basque', 'belarusian', 'bengali', 'catalan', 'chinese', 'danish', 'dutch', 'english', 'finnish', 'french', 'german', 'greek', 'hebrew', 'hinglish', 'hungarian', 'indonesian', 'italian', 'kazakh', 'nepali', 'norwegian', 'portuguese', 'romanian', 'russian', 'slovene', 'spanish', 'swedish', 'tajik', 'tamil', 'turkish']


In [6]:
print(stopwords.words('spanish'))

['de', 'la', 'que', 'el', 'en', 'y', 'a', 'los', 'del', 'se', 'las', 'por', 'un', 'para', 'con', 'no', 'una', 'su', 'al', 'lo', 'como', 'más', 'pero', 'sus', 'le', 'ya', 'o', 'este', 'sí', 'porque', 'esta', 'entre', 'cuando', 'muy', 'sin', 'sobre', 'también', 'me', 'hasta', 'hay', 'donde', 'quien', 'desde', 'todo', 'nos', 'durante', 'todos', 'uno', 'les', 'ni', 'contra', 'otros', 'ese', 'eso', 'ante', 'ellos', 'e', 'esto', 'mí', 'antes', 'algunos', 'qué', 'unos', 'yo', 'otro', 'otras', 'otra', 'él', 'tanto', 'esa', 'estos', 'mucho', 'quienes', 'nada', 'muchos', 'cual', 'poco', 'ella', 'estar', 'estas', 'algunas', 'algo', 'nosotros', 'mi', 'mis', 'tú', 'te', 'ti', 'tu', 'tus', 'ellas', 'nosotras', 'vosotros', 'vosotras', 'os', 'mío', 'mía', 'míos', 'mías', 'tuyo', 'tuya', 'tuyos', 'tuyas', 'suyo', 'suya', 'suyos', 'suyas', 'nuestro', 'nuestra', 'nuestros', 'nuestras', 'vuestro', 'vuestra', 'vuestros', 'vuestras', 'esos', 'esas', 'estoy', 'estás', 'está', 'estamos', 'estáis', 'están', 'e

In [7]:
text="""A wrapper around a sequence of simple (string) tokens, which is intended to support initial exploration of texts (via the interactive console).
Its methods perform a variety of analyses on the text’s contexts (e.g., counting, concordancing, collocation discovery), and display the results. If you wish to write a program which makes use of these analyses, then you should bypass the Text class, and use the appropriate analysis function or class directly instead.
"""

In [8]:
nltk.download('punkt_tab')

[nltk_data] Downloading package punkt_tab to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt_tab.zip.


True

# Tokenization


*   Break a Complex sentence into words
*   Understand the importance of each words with respect to the sentence
*  Produce a structural description on an input sentence




**Word Tokenization**
# word_tokenize():
This method used to break the given text into words.

In [9]:
tokens_NLTK=nltk.word_tokenize(text)

In [10]:
tokens_NLTK

['A',
 'wrapper',
 'around',
 'a',
 'sequence',
 'of',
 'simple',
 '(',
 'string',
 ')',
 'tokens',
 ',',
 'which',
 'is',
 'intended',
 'to',
 'support',
 'initial',
 'exploration',
 'of',
 'texts',
 '(',
 'via',
 'the',
 'interactive',
 'console',
 ')',
 '.',
 'Its',
 'methods',
 'perform',
 'a',
 'variety',
 'of',
 'analyses',
 'on',
 'the',
 'text',
 '’',
 's',
 'contexts',
 '(',
 'e.g.',
 ',',
 'counting',
 ',',
 'concordancing',
 ',',
 'collocation',
 'discovery',
 ')',
 ',',
 'and',
 'display',
 'the',
 'results',
 '.',
 'If',
 'you',
 'wish',
 'to',
 'write',
 'a',
 'program',
 'which',
 'makes',
 'use',
 'of',
 'these',
 'analyses',
 ',',
 'then',
 'you',
 'should',
 'bypass',
 'the',
 'Text',
 'class',
 ',',
 'and',
 'use',
 'the',
 'appropriate',
 'analysis',
 'function',
 'or',
 'class',
 'directly',
 'instead',
 '.']

In [11]:
len(tokens_NLTK)

90

**Sentence Tokenization**
# sent_tokenize():
This method used to break the given text into sentences

In [12]:
sentences_NLTK=nltk.sent_tokenize(text)
sentences_NLTK

['A wrapper around a sequence of simple (string) tokens, which is intended to support initial exploration of texts (via the interactive console).',
 'Its methods perform a variety of analyses on the text’s contexts (e.g., counting, concordancing, collocation discovery), and display the results.',
 'If you wish to write a program which makes use of these analyses, then you should bypass the Text class, and use the appropriate analysis function or class directly instead.']

In [13]:
len(sentences_NLTK)

3

**Stopword removal** is a process to remove common words from the document.



In [14]:
#remove Stopwords
filtered_tokens_NLTK=[word for word in tokens_NLTK if word.lower() not in stopwords.words('english')]
print("Original TOkens:",tokens_NLTK)
print("Filtered Tokens:",filtered_tokens_NLTK)
print("old text length:",len(tokens_NLTK))
print("new text length:",len(filtered_tokens_NLTK))

Original TOkens: ['A', 'wrapper', 'around', 'a', 'sequence', 'of', 'simple', '(', 'string', ')', 'tokens', ',', 'which', 'is', 'intended', 'to', 'support', 'initial', 'exploration', 'of', 'texts', '(', 'via', 'the', 'interactive', 'console', ')', '.', 'Its', 'methods', 'perform', 'a', 'variety', 'of', 'analyses', 'on', 'the', 'text', '’', 's', 'contexts', '(', 'e.g.', ',', 'counting', ',', 'concordancing', ',', 'collocation', 'discovery', ')', ',', 'and', 'display', 'the', 'results', '.', 'If', 'you', 'wish', 'to', 'write', 'a', 'program', 'which', 'makes', 'use', 'of', 'these', 'analyses', ',', 'then', 'you', 'should', 'bypass', 'the', 'Text', 'class', ',', 'and', 'use', 'the', 'appropriate', 'analysis', 'function', 'or', 'class', 'directly', 'instead', '.']
Filtered Tokens: ['wrapper', 'around', 'sequence', 'simple', '(', 'string', ')', 'tokens', ',', 'intended', 'support', 'initial', 'exploration', 'texts', '(', 'via', 'interactive', 'console', ')', '.', 'methods', 'perform', 'varie

In [15]:
s=stopwords.words('tamil')
print(list(s))

['அங்கு', 'அங்கே', 'அடுத்த', 'அதனால்', 'அதன்', 'அதற்கு', 'அதிக', 'அதில்', 'அது', 'அதே', 'அதை', 'அந்த', 'அந்தக்', 'அந்தப்', 'அன்று', 'அல்லது', 'அவன்', 'அவரது', 'அவர்', 'அவர்கள்', 'அவள்', 'அவை', 'ஆகிய', 'ஆகியோர்', 'ஆகும்', 'இங்கு', 'இங்கே', 'இடத்தில்', 'இடம்', 'இதனால்', 'இதனை', 'இதன்', 'இதற்கு', 'இதில்', 'இது', 'இதை', 'இந்த', 'இந்தக்', 'இந்தத்', 'இந்தப்', 'இன்னும்', 'இப்போது', 'இரு', 'இருக்கும்', 'இருந்த', 'இருந்தது', 'இருந்து', 'இவர்', 'இவை', 'உன்', 'உள்ள', 'உள்ளது', 'உள்ளன', 'எந்த', 'என', 'எனக்', 'எனக்கு', 'எனப்படும்', 'எனவும்', 'எனவே', 'எனினும்', 'எனும்', 'என்', 'என்ன', 'என்னும்', 'என்பது', 'என்பதை', 'என்ற', 'என்று', 'என்றும்', 'எல்லாம்', 'ஏன்', 'ஒரு', 'ஒரே', 'ஓர்', 'கொண்ட', 'கொண்டு', 'கொள்ள', 'சற்று', 'சிறு', 'சில', 'சேர்ந்த', 'தனது', 'தன்', 'தவிர', 'தான்', 'நான்', 'நாம்', 'நீ', 'பற்றி', 'பற்றிய', 'பல', 'பலரும்', 'பல்வேறு', 'பின்', 'பின்னர்', 'பிற', 'பிறகு', 'பெரும்', 'பேர்', 'போது', 'போன்ற', 'போல', 'போல்', 'மட்டுமே', 'மட்டும்', 'மற்ற', 'மற்றும்', 'மிக', 'மிகவும்', 'மீது', 'முதல்', '

In [16]:
s=stopwords.words('german')
print(list(s))

['aber', 'alle', 'allem', 'allen', 'aller', 'alles', 'als', 'also', 'am', 'an', 'ander', 'andere', 'anderem', 'anderen', 'anderer', 'anderes', 'anderm', 'andern', 'anderr', 'anders', 'auch', 'auf', 'aus', 'bei', 'bin', 'bis', 'bist', 'da', 'damit', 'dann', 'der', 'den', 'des', 'dem', 'die', 'das', 'dass', 'daß', 'derselbe', 'derselben', 'denselben', 'desselben', 'demselben', 'dieselbe', 'dieselben', 'dasselbe', 'dazu', 'dein', 'deine', 'deinem', 'deinen', 'deiner', 'deines', 'denn', 'derer', 'dessen', 'dich', 'dir', 'du', 'dies', 'diese', 'diesem', 'diesen', 'dieser', 'dieses', 'doch', 'dort', 'durch', 'ein', 'eine', 'einem', 'einen', 'einer', 'eines', 'einig', 'einige', 'einigem', 'einigen', 'einiger', 'einiges', 'einmal', 'er', 'ihn', 'ihm', 'es', 'etwas', 'euer', 'eure', 'eurem', 'euren', 'eurer', 'eures', 'für', 'gegen', 'gewesen', 'hab', 'habe', 'haben', 'hat', 'hatte', 'hatten', 'hier', 'hin', 'hinter', 'ich', 'mich', 'mir', 'ihr', 'ihre', 'ihrem', 'ihren', 'ihrer', 'ihres', 'euc

# Stemming
Stemming reduces works to their root by removing suffixes.



In [17]:
from nltk.stem import PorterStemmer

# create an object of class PorterStemmer
porter = PorterStemmer()
print(porter.stem("play"))
print(porter.stem("playing"))
print(porter.stem("plays"))
print(porter.stem("played"))

play
play
play
play


In [18]:
from nltk.stem import PorterStemmer
# create an object of class PorterStemmer
porter = PorterStemmer()
print(porter.stem("Communication"))

commun


In [19]:
from nltk.stem import PorterStemmer
# create an object of class PorterStemmer
porter = PorterStemmer()
print(porter.stem("vijayasrireddy"))

vijayasrireddi


# Lemmatization
 It reduces words to their base or root form.





In [20]:
from nltk.stem import WordNetLemmatizer
import nltk
# Download the 'wordnet' corpus
nltk.download('wordnet')


[nltk_data] Downloading package wordnet to /root/nltk_data...


True

In [21]:
from nltk.stem import WordNetLemmatizer

# create an object of class WordNetLemmatizer
lemmatizer = WordNetLemmatizer()
print(lemmatizer.lemmatize("plays", 'v'))
print(lemmatizer.lemmatize("played", 'v'))
print(lemmatizer.lemmatize("play", 'v'))
print(lemmatizer.lemmatize("playing", 'v'))

play
play
play
play


In [22]:
from nltk.stem import WordNetLemmatizer

# create an object of class WordNetLemmatizer
lemmatizer = WordNetLemmatizer()
print(lemmatizer.lemmatize("Communication"))


Communication


In [23]:
from nltk.stem import WordNetLemmatizer

# create an object of class WordNetLemmatizer
lemmatizer = WordNetLemmatizer()
print(lemmatizer.lemmatize("vijayasrireddy"))

vijayasrireddy


# Spacy
It is a advanced NLP in Python and used for handling large amounts of data.

In [24]:
import spacy
spacy_nlp=spacy.load('en_core_web_sm')

In [25]:

spacy_nlp=spacy.lang.en.stop_words.STOP_WORDS
print("List of stop words in english:\n%s"%list(spacy_nlp))

List of stop words in english:
['themselves', 'more', 'something', 'but', 'must', 'take', 'your', 'being', 'get', 'top', 'six', 'then', 'where', '‘re', 'might', 'move', '’m', 'whereafter', 'whatever', "'ve", 'does', 'third', 'make', 'from', 'before', 'several', 'latter', 'off', 'over', 'fifty', 'have', 'may', 'never', 'which', 'down', 'eleven', 'or', 'wherein', 'used', 'meanwhile', 'perhaps', 'an', 'nor', 'the', 'whereby', 'however', 'behind', 'first', 'via', 'both', 'neither', 'side', 'using', 'everyone', 'now', 'whither', 'anyone', 'put', 'yet', 'were', 'beside', 'until', 'hundred', 'various', 'on', 'n‘t', 'afterwards', 'except', 'seems', 'would', 'into', 'became', 'for', 'seemed', 'under', 'everywhere', 'of', 'up', 'somehow', 'through', 'four', 'one', 'formerly', 'thereby', 'no', 'three', 'noone', 'too', 'has', 'we', 'n’t', "'ll", 'always', 'without', 'though', 'regarding', 'nine', 'anywhere', 'indeed', 'since', 'two', 'eight', 'above', 'nobody', 'and', 'my', 'are', 'whereupon', 'tw

In [26]:
#length of stopwords
print("length of stop words",len(spacy_nlp))

length of stop words 326


In [27]:
#tokenization
import spacy
spacy_model = spacy.load('en_core_web_sm')

tokens_spacy = [token.text for token in spacy_model(text)]
print(tokens_spacy)


['A', 'wrapper', 'around', 'a', 'sequence', 'of', 'simple', '(', 'string', ')', 'tokens', ',', 'which', 'is', 'intended', 'to', 'support', 'initial', 'exploration', 'of', 'texts', '(', 'via', 'the', 'interactive', 'console', ')', '.', '\n', 'Its', 'methods', 'perform', 'a', 'variety', 'of', 'analyses', 'on', 'the', 'text', '’s', 'contexts', '(', 'e.g.', ',', 'counting', ',', 'concordancing', ',', 'collocation', 'discovery', ')', ',', 'and', 'display', 'the', 'results', '.', 'If', 'you', 'wish', 'to', 'write', 'a', 'program', 'which', 'makes', 'use', 'of', 'these', 'analyses', ',', 'then', 'you', 'should', 'bypass', 'the', 'Text', 'class', ',', 'and', 'use', 'the', 'appropriate', 'analysis', 'function', 'or', 'class', 'directly', 'instead', '.', '\n']


In [28]:
len(tokens_spacy)

91

In [29]:
filtered_tokens_spacy=[word for word in tokens_spacy if word.lower() not in spacy_nlp]
print("Original TOkens:",tokens_spacy)
print("Filtered Tokens:",filtered_tokens_spacy)
print("old text length:",len(tokens_spacy))
print("new text length:",len(filtered_tokens_spacy))

Original TOkens: ['A', 'wrapper', 'around', 'a', 'sequence', 'of', 'simple', '(', 'string', ')', 'tokens', ',', 'which', 'is', 'intended', 'to', 'support', 'initial', 'exploration', 'of', 'texts', '(', 'via', 'the', 'interactive', 'console', ')', '.', '\n', 'Its', 'methods', 'perform', 'a', 'variety', 'of', 'analyses', 'on', 'the', 'text', '’s', 'contexts', '(', 'e.g.', ',', 'counting', ',', 'concordancing', ',', 'collocation', 'discovery', ')', ',', 'and', 'display', 'the', 'results', '.', 'If', 'you', 'wish', 'to', 'write', 'a', 'program', 'which', 'makes', 'use', 'of', 'these', 'analyses', ',', 'then', 'you', 'should', 'bypass', 'the', 'Text', 'class', ',', 'and', 'use', 'the', 'appropriate', 'analysis', 'function', 'or', 'class', 'directly', 'instead', '.', '\n']
Filtered Tokens: ['wrapper', 'sequence', 'simple', '(', 'string', ')', 'tokens', ',', 'intended', 'support', 'initial', 'exploration', 'texts', '(', 'interactive', 'console', ')', '.', '\n', 'methods', 'perform', 'variety'

In [30]:
filtered_tokens_spacy

['wrapper',
 'sequence',
 'simple',
 '(',
 'string',
 ')',
 'tokens',
 ',',
 'intended',
 'support',
 'initial',
 'exploration',
 'texts',
 '(',
 'interactive',
 'console',
 ')',
 '.',
 '\n',
 'methods',
 'perform',
 'variety',
 'analyses',
 'text',
 'contexts',
 '(',
 'e.g.',
 ',',
 'counting',
 ',',
 'concordancing',
 ',',
 'collocation',
 'discovery',
 ')',
 ',',
 'display',
 'results',
 '.',
 'wish',
 'write',
 'program',
 'makes',
 'use',
 'analyses',
 ',',
 'bypass',
 'Text',
 'class',
 ',',
 'use',
 'appropriate',
 'analysis',
 'function',
 'class',
 'directly',
 'instead',
 '.',
 '\n']

In [31]:
print("existing length of the stopwords in nltk library:",len(nltk_stopwords))
print("existing length of the stopwords in spacy library:",len(spacy_nlp))

existing length of the stopwords in nltk library: 198
existing length of the stopwords in spacy library: 326


In [32]:
nltk_stopwords.append('example1')
len(nltk_stopwords)

199

In [33]:
nltk_stopwords.extend(['example2','example3'])
len(nltk_stopwords)

201

In [34]:
nltk_stopwords

['a',
 'about',
 'above',
 'after',
 'again',
 'against',
 'ain',
 'all',
 'am',
 'an',
 'and',
 'any',
 'are',
 'aren',
 "aren't",
 'as',
 'at',
 'be',
 'because',
 'been',
 'before',
 'being',
 'below',
 'between',
 'both',
 'but',
 'by',
 'can',
 'couldn',
 "couldn't",
 'd',
 'did',
 'didn',
 "didn't",
 'do',
 'does',
 'doesn',
 "doesn't",
 'doing',
 'don',
 "don't",
 'down',
 'during',
 'each',
 'few',
 'for',
 'from',
 'further',
 'had',
 'hadn',
 "hadn't",
 'has',
 'hasn',
 "hasn't",
 'have',
 'haven',
 "haven't",
 'having',
 'he',
 "he'd",
 "he'll",
 'her',
 'here',
 'hers',
 'herself',
 "he's",
 'him',
 'himself',
 'his',
 'how',
 'i',
 "i'd",
 'if',
 "i'll",
 "i'm",
 'in',
 'into',
 'is',
 'isn',
 "isn't",
 'it',
 "it'd",
 "it'll",
 "it's",
 'its',
 'itself',
 "i've",
 'just',
 'll',
 'm',
 'ma',
 'me',
 'mightn',
 "mightn't",
 'more',
 'most',
 'mustn',
 "mustn't",
 'my',
 'myself',
 'needn',
 "needn't",
 'no',
 'nor',
 'not',
 'now',
 'o',
 'of',
 'off',
 'on',
 'once',
 'on

In [35]:
nltk_stopwords.remove('example1')
len(nltk_stopwords)

200

In [36]:
nltk_stopwords

['a',
 'about',
 'above',
 'after',
 'again',
 'against',
 'ain',
 'all',
 'am',
 'an',
 'and',
 'any',
 'are',
 'aren',
 "aren't",
 'as',
 'at',
 'be',
 'because',
 'been',
 'before',
 'being',
 'below',
 'between',
 'both',
 'but',
 'by',
 'can',
 'couldn',
 "couldn't",
 'd',
 'did',
 'didn',
 "didn't",
 'do',
 'does',
 'doesn',
 "doesn't",
 'doing',
 'don',
 "don't",
 'down',
 'during',
 'each',
 'few',
 'for',
 'from',
 'further',
 'had',
 'hadn',
 "hadn't",
 'has',
 'hasn',
 "hasn't",
 'have',
 'haven',
 "haven't",
 'having',
 'he',
 "he'd",
 "he'll",
 'her',
 'here',
 'hers',
 'herself',
 "he's",
 'him',
 'himself',
 'his',
 'how',
 'i',
 "i'd",
 'if',
 "i'll",
 "i'm",
 'in',
 'into',
 'is',
 'isn',
 "isn't",
 'it',
 "it'd",
 "it'll",
 "it's",
 'its',
 'itself',
 "i've",
 'just',
 'll',
 'm',
 'ma',
 'me',
 'mightn',
 "mightn't",
 'more',
 'most',
 'mustn',
 "mustn't",
 'my',
 'myself',
 'needn',
 "needn't",
 'no',
 'nor',
 'not',
 'now',
 'o',
 'of',
 'off',
 'on',
 'once',
 'on

# Difference between NLTK and Spacy
**NLTK**
* NLTK is mainly a string processing library.

* Provides access to many algorithms.If you care about specific algorithms and customizations go with NLTK.
* NLTK is old library.User community as active as Spacy.
* 198 Stopwords.


**SPACY**
* Spacy is Object oriented.
* Provides most efficient NLP algortihm for a given task.Hence if you care about the end result,go with Spacy.
* Spacy is new library and has a very active user community.
* 326 Stopwords.

In [37]:
import spacy

In [38]:
nlp=spacy.load("en_core_web_sm")
doc=nlp("Vijaya sri Manda from Seshadri Rao gudlavalleru Engineering College.Currently pursuing 3rd year btech in Artificial Intelligence and Data science")
for sentence in doc.sents:
  print(sentence)

Vijaya sri Manda from Seshadri Rao gudlavalleru Engineering College.
Currently pursuing 3rd year btech in Artificial Intelligence and Data science


In [39]:
for sentence in doc.sents:
  for word in sentence:
    print(word)

Vijaya
sri
Manda
from
Seshadri
Rao
gudlavalleru
Engineering
College
.
Currently
pursuing
3rd
year
btech
in
Artificial
Intelligence
and
Data
science


In [40]:
from nltk.tokenize import sent_tokenize

In [41]:
sent_tokenize("Vijaya sri Manda from Seshadri Rao gudlavalleru Engineering College.Currently pursuing 3rd year btech in Artificial Intelligence and Data science")

['Vijaya sri Manda from Seshadri Rao gudlavalleru Engineering College.Currently pursuing 3rd year btech in Artificial Intelligence and Data science']

In [42]:
from nltk.tokenize import word_tokenize
word_tokenize("Vijaya sri Manda from Seshadri Rao gudlavalleru Engineering College.Currently pursuing 3rd year btech in Artificial Intelligence and Data science")

['Vijaya',
 'sri',
 'Manda',
 'from',
 'Seshadri',
 'Rao',
 'gudlavalleru',
 'Engineering',
 'College.Currently',
 'pursuing',
 '3rd',
 'year',
 'btech',
 'in',
 'Artificial',
 'Intelligence',
 'and',
 'Data',
 'science']

In [43]:
# customation of stopwords
text='''The community of volunteers behind Wikipedia is the most important and unique element of Wikipedia’s success. For nearly 25 years, Wikipedia editors have researched, deliberated, discussed, built consensus, and collaboratively written the largest encyclopedia humankind has ever seen. Their care and commitment to reliable encyclopedic knowledge is something AI cannot replace.

That is why our new AI strategy doubles down on the volunteers behind Wikipedia.

We will use AI to build features that remove technical barriers to allow the humans at the core of Wikipedia to spend their valuable time on what they want to accomplish, and not on how to technically achieve it. Our investments will be focused on specific areas where generative AI excels, all in the service of creating unique opportunities that will boost Wikipedia’s volunteers:

'''

In [44]:
custom_stopwords=['Wikipedia','AI','the','to','is']

In [45]:
tokens=nltk.word_tokenize(text)

In [46]:
print(tokens)

['The', 'community', 'of', 'volunteers', 'behind', 'Wikipedia', 'is', 'the', 'most', 'important', 'and', 'unique', 'element', 'of', 'Wikipedia', '’', 's', 'success', '.', 'For', 'nearly', '25', 'years', ',', 'Wikipedia', 'editors', 'have', 'researched', ',', 'deliberated', ',', 'discussed', ',', 'built', 'consensus', ',', 'and', 'collaboratively', 'written', 'the', 'largest', 'encyclopedia', 'humankind', 'has', 'ever', 'seen', '.', 'Their', 'care', 'and', 'commitment', 'to', 'reliable', 'encyclopedic', 'knowledge', 'is', 'something', 'AI', 'can', 'not', 'replace', '.', 'That', 'is', 'why', 'our', 'new', 'AI', 'strategy', 'doubles', 'down', 'on', 'the', 'volunteers', 'behind', 'Wikipedia', '.', 'We', 'will', 'use', 'AI', 'to', 'build', 'features', 'that', 'remove', 'technical', 'barriers', 'to', 'allow', 'the', 'humans', 'at', 'the', 'core', 'of', 'Wikipedia', 'to', 'spend', 'their', 'valuable', 'time', 'on', 'what', 'they', 'want', 'to', 'accomplish', ',', 'and', 'not', 'on', 'how', 't

In [47]:
filtered_tokens=[word for word in tokens if word.lower() not in custom_stopwords]
print(filtered_tokens)

['community', 'of', 'volunteers', 'behind', 'Wikipedia', 'most', 'important', 'and', 'unique', 'element', 'of', 'Wikipedia', '’', 's', 'success', '.', 'For', 'nearly', '25', 'years', ',', 'Wikipedia', 'editors', 'have', 'researched', ',', 'deliberated', ',', 'discussed', ',', 'built', 'consensus', ',', 'and', 'collaboratively', 'written', 'largest', 'encyclopedia', 'humankind', 'has', 'ever', 'seen', '.', 'Their', 'care', 'and', 'commitment', 'reliable', 'encyclopedic', 'knowledge', 'something', 'AI', 'can', 'not', 'replace', '.', 'That', 'why', 'our', 'new', 'AI', 'strategy', 'doubles', 'down', 'on', 'volunteers', 'behind', 'Wikipedia', '.', 'We', 'will', 'use', 'AI', 'build', 'features', 'that', 'remove', 'technical', 'barriers', 'allow', 'humans', 'at', 'core', 'of', 'Wikipedia', 'spend', 'their', 'valuable', 'time', 'on', 'what', 'they', 'want', 'accomplish', ',', 'and', 'not', 'on', 'how', 'technically', 'achieve', 'it', '.', 'Our', 'investments', 'will', 'be', 'focused', 'on', 's

In [48]:
len(tokens)

147

In [49]:
len(filtered_tokens)

131